integrations/go/workload_executor_test.go

```diff
@@ -0,0 +1,158 @@
+		if err := mtest.Teardown(); err != nil {
+			panic(err)
+		}
+	}()
+
+	// killAllSessions will return an auth error if it's run
+	fileReqs, runners := unified.ParseTestFile(t, workloadSpec, unified.NewOptions().SetRunKillAllSessions(false))
+
+	mtOpts := mtest.NewOptions().
+		RunOn(fileReqs...).
+		CreateClient(false)
+	mt := mtest.New(t, mtOpts)
+	defer mt.Close()
+
+	// a testfile can parse to multiple runners, though we only expect to get one per file
+	for _, runner := range runners {
+		mtOpts := mtest.NewOptions().
+			RunOn(runner.RunOnRequirements...).
+			CreateClient(false)
+
+		mt.RunOpts(runner.Description, mtOpts, func(mt *mtest.T) {
+			// the workload executor should be able to run non-looping tests and EndLoop() will panic
+			// if the test has already finished
+			if hasLoop(runner) {
+				// Waits for the termination signal from astrolabe and terminates the loop operation
+				go func() {
+					c := make(chan os.Signal, 1)
+					signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+
+					<-c
+					runner.EndLoop()
+				}()
+			}
+
+			testErr := runner.Run(mt)
+
+			em := runner.GetEntities()
+
+			// structs for marshaling json results
+			results := struct {
+				NumErrors     int   `bson:"numErrors"`
+				NumFailures   int   `bson:"numFailures"`
+				NumSuccesses  int32 `bson:"numSuccesses"`
+				NumIterations int32 `bson:"numIterations"`
+			}{}
+			allEvents := struct {
+				Events   []bson.Raw `bson:"events"`
+				Errors   []bson.Raw `bson:"errors"`
+				Failures []bson.Raw `bson:"failures"`
+			}{}
+
+			var err error
+			if results.NumIterations, err = em.Iterations("iterations"); err != nil {
+				results.NumIterations = -1
+			}
+			if results.NumSuccesses, err = em.Successes("successes"); err != nil {
+				results.NumSuccesses = -1
+			}
+
+			allEvents.Failures, _ = em.BSONArray("failures")
+			allEvents.Errors, _ = em.BSONArray("errors")
+			allEvents.Events, _ = em.EventList("events")
+			// a non-nil testErr should be added to the appropriate slice
+			if testErr != nil {
+				errDoc := bson.Raw(bsoncore.NewDocumentBuilder().
+					AppendString("error", testErr.Error()).
+					AppendInt64("time", time.Now().Unix()).
+					Build())
+				switch {
+				// check for the failure substring
+				case strings.Contains(testErr.Error(), " verification failed:"):
+					allEvents.Failures = append(allEvents.Failures, errDoc)
+				default:
+					allEvents.Errors = append(allEvents.Errors, errDoc)
+				}
+			}
+
+			results.NumErrors = len(allEvents.Errors)
+			results.NumFailures = len(allEvents.Failures)
+
+			// make sure that empty slices marshal as slices instead of null
+			if len(allEvents.Events) == 0 {
+				allEvents.Events = make([]bson.Raw, 0)
+			}
+			if results.NumErrors == 0 {
+				allEvents.Errors = make([]bson.Raw, 0)
+			}
+			if results.NumFailures == 0 {
+				allEvents.Failures = make([]bson.Raw, 0)
+			}
+
+			path, _ := os.Getwd()
+
+			resultJSON, err := bson.MarshalExtJSON(&results, false, false)
```

Comment:
Two questions:

1. Why `bson.MarshalExtJSON` rather than `json.Marshal`? I see no mention of ext JSON in the executor spec.
2. Why pass a pointer to `results` rather than the value?

Reply:
gone away